### PR TITLE
Add protocol option to all queries

### PR DESCRIPTION
**Pull request by fsuste**
## Pull Request Checklist

**Is this in reference to an existing issue?**

#### General

- [x] Update Changelog following the conventions laid out on [Keep A Changelog](http://keepachangelog.com/)

- [x] Update README with any necessary configuration snippets

- [x] Binstubs are created if needed

- [x] RuboCop passes

- [x] Existing tests pass

#### New Plugins

- [x] Tests

- [x] Add the plugin to the README

- [x] Does it have a complete header as outlined [here](http://sensu-plugins.io/docs/developer_guidelines.html#coding-style)

### CODE REVIEW

**Comment by fsuste** [comment]
This is a pull request resolving issue #8

**Comment by majormoses** [comment]
Thanks for your contribution to Sensu plugins! Without people like you submitting PRs we couldn't run the project. I will review it shortly.

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,8 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 This CHANGELOG follows the format listed at [Keep A Changelog](http://keepachangelog.com/)
 
 ## [Unreleased]
+### Changed
```
Comment (majormoses) [inline CHANGELOG.md]: This should be `### Added` as that is a new feature the ability to change the protocol.

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,8 @@ This project adheres to [Semantic Versioning](http://semver.org/).
...
+- Add protocol option which can be passed asn an argument to specify which protocol to use
```
Comment (majormoses) [inline CHANGELOG.md]: > asn an argument

looks like a typo

File: bin/check-graylog-buffers.rb [hunk 3]
```diff
@@ -37,6 +37,12 @@
 require 'rest-client'
 
 class CheckGraylogBuffers < Sensu::Plugin::Check::CLI
+  option :protocol,
+         description: 'Protocol for connecting to Graylog',
+         short: '-proto',
```
Comment (majormoses) [inline bin/check-graylog-buffers.rb]: I am actually not sure if short is ok with multi char short args. Have you tested this? Usually short options single chars that can be chained together like `tar -xvf` that is not a single arg but actually 3 of them.

File: bin/check-graylog-buffers.rb [hunk 4]
```diff
@@ -37,6 +37,12 @@
...
+         long: '--protocol PROTOCOL',
+         default: 'http'
```
Comment (majormoses) [inline bin/check-graylog-buffers.rb]: As much as I feel `https` should be the default I agree that the breaking change would be unwarranted right now.

**Review by majormoses — CHANGES_REQUESTED** [review]
Overall this looks great one minor changelog tweak and a [testing artifact](https://github.com/sensu-plugins/community/blob/master/PULL_REQUEST_PROCESS.md#7-testing-artifacts) is all I need and to get this merged and released.

**Comment by fsuste** [comment]
I would also like to see `https` as a default, but that is an expensive change version wise... I am now working on a testing artifact.

**Comment by fsuste** [comment]
I don't actually see how can I test this except for unit testing, but that would be testing the mixlib-cli gem... If you have a better idea for this, I would love to hear it :)

**Comment by majormoses** [comment]
Well there are a couple options:
- unit testing
- integration testing
- manual testing

I would suggest you read the following articles where I outline the types of testing and go in depth with integration testing:
- https://blog.sensuapp.org/our-approach-to-testing-sensu-plugins-869260eba94c
- https://blog.sensuapp.org/writing-sensu-plugin-tests-with-test-kitchen-and-serverspec-b646d2eeee51

The easiest is a manual test, just run the check you would have run via sensu in cli and capture the input and output in a gist.

**Comment by nathanhruby** [comment]
The tests in check-graylog-streams_spec.rb could be used as an example for checking the args and stubbing out the actual call to make sure it is used as intended.

File: bin/check-graylog2-alive.rb [hunk 5]
```diff
@@ -37,6 +37,12 @@
 require 'rest-client'
 
 class CheckGraylog2Alive < Sensu::Plugin::Check::CLI
+  option :protocol,
+         description: 'Protocol for connecting to Graylog',
+         short: '-s',
```
Comment (majormoses) [inline bin/check-graylog2-alive.rb]: let's drop the `-s` as it no longer makes sense and `-p` is usually `port`

File: bin/check-graylog-streams.rb [hunk 6]
```diff
@@ -26,6 +26,12 @@
 require 'json'
 
 class CheckGraylogStreams < Sensu::Plugin::Check::CLI
+  option :protocol,
+         description: 'Protocol for connecting to Graylog',
+         short: '-s',
```
Comment (majormoses) [inline bin/check-graylog-streams.rb]: let's drop the `-s` as it no longer makes sense and `-p` is usually `port`

File: bin/check-graylog-buffers.rb [hunk 7]
```diff
@@ -37,6 +37,12 @@
 require 'rest-client'
 
 class CheckGraylogBuffers < Sensu::Plugin::Check::CLI
+  option :protocol,
+         description: 'Protocol for connecting to Graylog',
+         short: '-s',
```
Comment (majormoses) [inline bin/check-graylog-buffers.rb]: let's drop the `-s` as it no longer makes sense and `-p` is usually `port`

File: bin/metrics-graylog.rb [hunk 8]
```diff
@@ -38,6 +38,12 @@
 require 'rest-client'
 
 class MetricsGraylog < Sensu::Plugin::Metric::CLI::Graphite
+  option :protocol,
+         description: 'Protocol for connecting to Graylog',
+         short: '-s',
```
Comment (majormoses) [inline bin/metrics-graylog.rb]: let's drop the `-s` as it no longer makes sense and `-p` is usually `port`

**Review by majormoses — CHANGES_REQUESTED** [review]
A few more things I noticed that could be improved:
- drop the `-s` option as the name does not make sense and `-p` is used for `port`
- specify the available options for `protocol` like [this](https://github.com/sensu-plugins/sensu-plugins-redis/blob/2.3.2/lib/redis_client_options.rb#L75)

**Comment by majormoses** [comment]
I have made the self requested modifications if someone can provide either a manual or automated testing artifact I am :+1: to merge this.

**Comment by majormoses** [comment]
@nathanhruby @Narayanan170  any chance you have a graylog that we can test against? I don't see how this PR is the culprit of https://github.com/sensu-plugins/sensu-plugins-graylog/issues/8#issuecomment-365294851 but it does mean we need to ensure that we have a test before I am OK with merging this.

**Comment by nathanhruby** [comment]
I don't work at a place with graylog anymore, but I might have some time later this week to spin up a test cluster in a docker as test of something else I am doing.

**Comment by majormoses** [comment]
I can confirm this works:
```
$ bundle exec ./bin/check-graylog2-alive.rb --protocol https --host graylog.$DOMAIN.$TLD --username babrams --password $(cat creds) --port 443 --apipath /api
CheckGraylog2Alive OK: Graylog2 server is: running/true/alive
```

**Comment by majormoses** [comment]
released: https://rubygems.org/gems/sensu-plugins-graylog/versions/1.3.0

**Pull request merged**